src/hooks/cgroups/pbs_cgroups.PY

```diff
@@ -0,0 +1,3657 @@
+                key = entries[0].strip()
+                val = entries[1].strip()
+                if proc is None and key != 'processor':
+                    raise ProcessingError('Failed to parse /proc/cpuinfo')
+                if key == 'processor':
+                    proc = int(val)
+                    if proc in cpuinfo:
+                        raise ProcessingError('Duplicate CPU ID found')
+                    cpuinfo['cpu'][proc] = {}
+                    cpuinfo['cpu'][proc]['threads'] = []
+                elif key == 'flags':
+                    cpuinfo['cpu'][proc][key] = val.split()
+                elif val.isdigit():
+                    cpuinfo['cpu'][proc][key] = int(val)
+                else:
+                    cpuinfo['cpu'][proc][key] = val
+        if not cpuinfo['cpu']:
+            raise ProcessingError('No CPU information found')
+        cpuinfo['logical_cpus'] = len(cpuinfo['cpu'].keys())
+        cpuinfo['hyperthreads_per_core'] = 1
+        cpuinfo['hyperthreads'] = []
+        try:
+            if 'Intel' in cpuinfo['cpu'][0]['vendor_id']:
+                if 'ht' in cpuinfo['cpu'][0]['flags']:
+                    cpuinfo['hyperthreads_per_core'] = \
+                        cpuinfo['cpu'][0]['siblings'] / \
+                        cpuinfo['cpu'][0]['cpu cores']
+                    # Map hyperthreads to physical cores
+                    if cpuinfo['hyperthreads_per_core'] > 1:
+                        pbs.logmsg(pbs.EVENT_DEBUG3,
+                                   "Mapping hyperthreads to cores")
+                        cores = cpuinfo['cpu'].keys()
+                        threads = []
+                        for xid in cores:
+                            x = cpuinfo['cpu'][xid]
+                            for yid in cores:
+                                y = cpuinfo['cpu'][yid]
+                                if xid >= yid:
+                                    continue
+                                if x['physical id'] != y['physical id']:
+                                    continue
+                                if x['core id'] == y['core id']:
+                                    cpuinfo['cpu'][xid]['threads'].append(yid)
+                                    threads.append(yid)
+                        pbs.logmsg(pbs.EVENT_DEBUG3, "HT cores: %s" % threads)
+                        cpuinfo['hyperthreads'] = threads
+        except:
+            pbs.logmsg(pbs.EVENT_DEBUG, "%s: Hyperthreading check failed" %
+                       (caller_name()))
+        cpuinfo['physical_cpus'] = cpuinfo['logical_cpus'] / \
+            cpuinfo['hyperthreads_per_core']
+        return cpuinfo
+
+    # Gather the jobs running on this node
+    def gather_jobs_on_node(self):
+        pbs.logmsg(pbs.EVENT_DEBUG3, "%s: Method called" % (caller_name()))
+        # Get the job list from the server
+        jobs = None
+        try:
+            jobs = pbs.server().vnode(self.hostname).jobs
```

Comment:
This was a very good catch. Querying the server is not scalable. The configure_job method was updated to call gather_jobs_on_node which no longer queries the server.